File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    let link=new Map();
    for(let [from, to] of tickets){
        if(!link.has(from)) link.set(from,[]);
        link.get(from).push(to);
    }
    for(let list of link.values()){
        list.sort();
    }
    
    let answer;
    bt('ICN',['ICN']);
    function bt(curr, route){
        if(route.length===tickets.length+1){
            answer=[...route];
            return true;
        }
        let list=link.get(curr);
        if(!link.has(curr) || list.length===0) return false;
        
        for(let i=0; i<list.length; i++){
            const next=list[i];
            list.splice(i,1);
            route.push(next);
            if(bt(next, route)) return true;
            route.pop();
            list.splice(i, 0, next);
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    class Pair{
        String key;
        int idx;

        Pair(String key, int idx){
            this.key = key;
            this.idx = idx;
        }
    }
    Map<String, List<Pair>> map = new HashMap<>();
    String[] answer;
    public String[] solution(String[][] tickets) {
        int len = tickets.length;
        
        for(int i=0; i<len; i++){
            String[] ticket = tickets[i];
            map.computeIfAbsent(ticket[0], v-> new ArrayList<>()).add(new Pair(ticket[1], i));
        }
        
        for(List<Pair> list : map.values()) {
            list.sort((a,b)-> a.key.compareTo(b.key));
        }
        List<String> route = new ArrayList<>();
        route.add("ICN");
        bt("ICN", route, new boolean[len]);
        
        return answer;
    }
    
    public void bt(String cur, List<String> route, boolean[] visited){
        if(answer != null) return;
        
        if(route.size()>=visited.length+1){
            answer = route.toArray(new String[0]);
            return;
        }
        
        List<Pair> nexts = map.get(cur);
        if(nexts == null) return;
        for(Pair next: nexts){
            if(visited[next.idx]==false){
                visited[next.idx]=true;
                route.add(next.key);
                bt(next.key, route, visited);
                route.remove(route.size()-1);
                visited[next.idx]=false;
            }
        }
        
    }
}